google/cloud/security/scanner/audit/instance_network_interface_rules_engine.py

```diff
@@ -0,0 +1,251 @@
+            Assigned automatically when the rule book is built.
+        """
+        project = rule_def.get('project')
+        network = rule_def.get('network')
+        whitelist = rule_def.get('whitelist')
+        is_external_network = rule_def.get('is_external_network')
+
+        if (whitelist is None) or (project is None) or (network is None)\
+                or (is_external_network is None):
+            raise audit_errors.InvalidRulesSchemaError(
+                'Faulty rule {}'.format(rule_def.get('name')))
+
+        rule_def_resource = {'whitelist': whitelist,
+                             'project': escape_and_globify(project),
+                             'network': escape_and_globify(network),
+                             'is_external_network': is_external_network}
+
+        rule = Rule(rule_name=rule_def.get('name'),
+                    rule_index=rule_index,
+                    rules=rule_def_resource)
+
+        resource_rules = self.resource_rules_map.get(rule_index)
+        if not resource_rules:
+            self.resource_rules_map[rule_index] = rule
+
+    def get_resource_rules(self):
+        """Get all the resource rules for (resource, RuleAppliesTo.*).
+
+        Returns:
+            A list of ResourceRules.
+        """
+        return self.resource_rules_map.values()
+
+
+class Rule(object):
+    """The rules class for instance_network_interface."""
+
+    def __init__(self, rule_name, rule_index, rules):
+        """Initialize.
+
+        Args:
+            rule_name: Name of the loaded rule
+            rule_index: The index of the rule from the  definitions
+            rules: the network instance
+            the resources to accosiate with the rules
+             (so far every network_instance)
+        """
+        self.rule_name = rule_name
+        self.rule_index = rule_index
+        self.rules = rules
+
+    def find_violations(self, instance_network_interface):
+        """
+        Raise violation is the ip is not in the whitelist.
+
+        Args:
+                instance_network_interface: InstanceNetworkInterface obj
+        """
+        network_and_project = re.search(
+            'compute\/v1\/projects\/([^\/]*).*networks\/([^\/]*)',
+            instance_network_interface.network)
+        project = network_and_project.group(1)
+        network = network_and_project.group(2)
+        is_external_network = instance_network_interface.access_configs \
+            is not None
+        if not self.rules['whitelist'].get(project):
+            if is_external_network:
+                yield self.RuleViolation(
+                    resource_type='instance',
+                    rule_name=self.rule_name,
+                    rule_index=self.rule_index,
+                    violation_type='UNENFORCED_NETWORK_VIOLATION',
+                    project=project,
+                    network=network,
+                    ips='not an ip')
```

Comment:
I noticed that in your RuleViolation definition, in the comments you have "ip", which is a string, but the namedtuple property is "ips". If you are tracking just a single IP, could you change the namedtuple property to "ip"? I also noticed the scanner "flattening scheme", you are also setting a property to "ip". So it would be great to make this consistent.

Also, would it be cleaner to treat the "not an ip" case as just "None", and handle that upstream?

Let me know if this is not clear.